Sources/PostgreSQL/Connection.swift

```diff
@@ -118,18 +119,63 @@ public final class Connection: ConnectionProtocol {
     }
 
     public func open() throws {
-        connection = PQsetdbLogin(
-            connectionInfo.host,
-            String(connectionInfo.port),
-            connectionInfo.options ?? "",
-            connectionInfo.tty ?? "",
-            connectionInfo.databaseName,
-            connectionInfo.username ?? "",
-            connectionInfo.password ?? ""
-        )
-
-        if let error = mostRecentError {
-            throw error
+        guard connection == nil else {
+            throw ConnectionError(description: "Connection already opened.")
+        }
+        
+        var components = URLComponents()
+        components.scheme = "postgres"
+        components.host = connectionInfo.host
+        components.port = connectionInfo.port
+        components.user = connectionInfo.username
+        components.password = connectionInfo.password
+        components.path = "/\(connectionInfo.databaseName)"
+        if let options = connectionInfo.options {
+            components.queryItems = [URLQueryItem(name: "options", value: options)]
+        }
+        let url = components.url!.absoluteString
+        
+        connection = PQconnectStart(url)
+        
+        guard connection != nil else {
+            throw ConnectionError(description: "Could not allocate connection.")
+        }
+        
+        guard PQstatus(connection) != CONNECTION_BAD else {
+            throw ConnectionError(description: "Could not start connection.")
+        }
+        
+        fd = PQsocket(connection)
+        guard fd >= 0 else {
+            throw mostRecentError ?? ConnectionError(description: "Could not get file descriptor.")
+        }
+
+        loop: while true {
+            let status = PQconnectPoll(connection)
+            switch status {
+            case PGRES_POLLING_OK:
+                break loop
+            case PGRES_POLLING_READING:
+                mill_fdwait_(fd, FDW_IN, 15.seconds.fromNow().int64milliseconds, nil)
+                mill_fdclean_(fd)
+            case PGRES_POLLING_WRITING:
+                mill_fdwait_(fd, FDW_OUT, 15.seconds.fromNow().int64milliseconds, nil)
+                mill_fdclean_(fd)
+            case PGRES_POLLING_ACTIVE:
+                break
```

Comment:
`continue` instead of `break` to avoid the label. This might be "clever", though...